lib/elixir/pages/Library Guidelines.md

```diff
@@ -0,0 +1,214 @@
+rescue
+  File.Error ->
+    :it_failed
+end
+```
+
+you should prefer:
+
+```elixir
+case File.read("some_path_that_may_or_may_not_exist") do
+  {:ok, contents} -> {:it_worked, contents}
+  {:error, _} -> :it_failed
+end
+```
+
+As a library author, it is your responsibility to make sure users are not required to use exceptions for control-flow in their applications. You can follow the same convention as Elixir here, using function names without `!` for returning `:ok`/`:error` tuples and appending `!` to those function names for the raising version.
+
+It is important to note that a name without `!` does not mean a function will never raise. For example, even `File.read/1` can fail in case of bad arguments:
+
+```iex
+iex> File.read(1)
+** (FunctionClauseError) no function clause matching in IO.chardata_to_string/1
+```
+
+The usage of `:ok`/`:error` tuples is about the domain that the function works on, in this case, filesystem access. Bad arguments, logical errors, invalid options should raise regardless of the function name. 
+
+### Avoid application configuration
+
+You should avoid using [the application environment](https://hexdocs.pm/elixir/Application.html#get_env/2) for configuring your own libraries. The application environment is **global** which means it becomes impossible for two dependencies to use your library in two different ways.
+
+Let's see a simple example. Imagine that you implement a library that breaks a string in two parts based on the first occurrence of the dash `-` character:
+
+```elixir
+defmodule DashSplitter do
+  def split(string) when is_binary(string) do
+    String.split(string, "-", parts: 2)
+  end
+end
+```
+
+Now imagine someone wants to split the string in three parts. You decide to make the number of parts configurable via the application environment:
+
+```elixir
+def split(string) when is_binary(string) do
+  parts = Application.get_env(:dash_splitter, :parts, 2)
+  String.split(string, "-", parts: parts)
+end
+```
+
+Now users can configure your library in their `config/config.exs` file as follows:
+
+```elixir
+config :dash_splitter, :parts, 3
+```
+
+Once your library is configured, it will change the behaviour of all users of your library. If a library was expecting it to split the string in 2 parts, since the configuration is global, it will now split it in 3 parts.
+
+The solution is to provide configuration as close as possible to where it is used and not via the application environment. In case of a function, you could use a keyword list:
+
+```elixir
+def split(string, opts \\ []) when is_binary(string) and is_list(opts) do
+  parts = Keyword.get(opts, :parts, 2)
+  String.split(string, "-", parts: parts)
+end
+```
+
+In case you need to configure a process, the options should be passed when starting that process. 
+
+The application environment should be reserved only for configurations that are truly global, for example, to control your application boot process and its supervision tree.
```

Comment:
Question: What about defaults? I.e. for wallaby or HTTP libraries the wait timeout is ok to my mind to configure on the application level as long as individual callers can still override it.

To bring up another example, I've been thinking to allow configuring the default formatters in benchee (for instance) - for people who always want to use the HTML formatter for instance.

Are the mentioned use cases still bad or are they an exception from the rule worth documenting?

(I know explicit is better than implicit but especially in the timeout case I feel like it'd be an advantage)

Reply:
The problem with defaults is that you still change everybody's defaults when you change the default value. Imagine I write some HTTP code with the assumption that the timeout will be five seconds and decide to rely on the default, that assumption may not be true, and the best would be for me to explicitly pass 5 seconds then.